test/storagelib.py

```diff
@@ -0,0 +1,210 @@
+
+    def inode(self, f):
+        return self.machine.execute("stat -L '%s' -c %%i" % f)
+
+    # Action buttons
+
+    def content_action_btn(self, index):
+        return "#content .list-group li:nth-child(%d) .btn-group" % index
+
+    def content_action(self, index, action):
+        self.browser.click_action_btn (self.content_action_btn(index), action)
+
+    def content_default_action(self, index, action):
+        btn = self.content_action_btn(index)
+        self.browser.wait_action_btn (btn, action)
+        self.browser.click_action_btn (btn)
+
+    def content_single_action(self, index, action):
+        btn = "#content .list-group li:nth-child(%d) button" % index
+        self.browser.wait_text (btn, action)
+        self.browser.click (btn)
+
+    # Dialogs
+
+    def dialog_wait_open(self):
+        self.browser.wait_present('#dialog')
+        self.browser.wait_visible('#dialog')
+
+    def dialog_wait_alert(self, text):
+        self.browser.wait_in_text('#dialog .alert-message', text)
+
+    def dialog_field(self, field):
+        return '#dialog [data-field="%s"]' % field
+
+    def dialog_set_val(self, field, val):
+        if isinstance(val, bool):
+            self.browser.set_checked(self.dialog_field(field), val)
+        else:
+            self.browser.set_val(self.dialog_field(field), val)
+
+    def dialog_select(self, field, label, val):
+        self.browser.set_checked('%s .checkbox:contains("%s") input' % (self.dialog_field(field), label), val)
+
+    def dialog_wait_val(self, field, val):
+        self.browser.wait_val(self.dialog_field(field), val)
+
+    def dialog_wait_error(self, field, val):
+        # XXX - allow for more than one error
+        self.browser.wait_in_text('#dialog .dialog-error', val)
+
+    def dialog_wait_not_visible(self, field):
+        self.browser.wait_not_visible(self.dialog_field(field))
+
+    def dialog_apply(self):
+        self.browser.click('#dialog [data-action="apply"]')
+
+    def dialog_cancel(self):
+        self.browser.click('#dialog [data-dismiss="modal"]')
+
+    def dialog_wait_close(self):
+        self.browser.wait_not_present('#dialog')
+
+    def dialog(self, values, expect={}):
+        self.dialog_wait_open()
+        for f in expect:
+            self.dialog_wait_val(f, expect[f])
+        for f in values:
+            self.dialog_set_val(f, values[f])
+        self.dialog_apply()
+        self.dialog_wait_close()
+
+    def confirm(self):
+        self.dialog({})
+
+    # Deep inspection
+
+    # The devices.js scope allows us to execute code inside of it.  We
+    # use that to wait for the internal state to reach an expected
+    # point before opening dialogs that depend on that state.
+
+    def wait_devices_func(self, code, *args):
+        self.browser.wait_js_func("""function (code, args) {
+            return window.phantom_eval_in_devices("(" + code + ")").apply(this, args);
```

Comment:
This seems quite hacky. Shouldn't we just tag the displayed tags with semantic attributes, and not need to have this extremely fragile code poking around inside the internals of the storage page?